### ISSUE TITLE
TranscriptLoggerMiddleware performance regression 

### ISSUE THREAD

**Issue by tomlm** [microsoft/botbuilder-dotnet#4642]
## Version

4.10.3

## Describe the bug

With commit Commit 6720f2c we introduced a performance regression in the transcript logger.

The transcript logger was designed to be zero impact which it did by putting all activities to be logged into a buffer queue. 

At the end of the turn, it drained the queue as async tasks, but it did NOT await those tasks so that they were fire and forget background tasks.  

The performance regression was introduced because Stylecop complaining about the the non-await on an Tasks. Even though there was a comment saying we were deliberately not awaiting these tasks, it was replaced with code which waits for all of the tasks to complete.  Awaiting for this blocks the completion of the turn.  

This performance blocking of awaiting causes a cascade of bottlenecks in the entire system.  

1.  Bot Log(Activity1) => Skill  Log(Activity1)
2.  Skill Sends Trace Activity Log(Activity2) => Bot logs incoming activity Log(Activity2)
3.  Because activities are routed back out, Bot Sends incoming activity to channel which logs it again (LogActivity(Activity2v2))
4.  Skill Sends Trace Activity Log(Activity3) => Bot logs incoming activity Log(Activity3)
5.  Because activities are routed back out, Bot Sends incoming activity to channel which logs it again (LogActivity(Activity3v2))
6.  Skill Sends Trace Activity Log(Activity4) => Bot logs incoming activity Log(Activity4)
7.  Because activities are routed back out, Bot Sends incoming activity to channel which logs it again (LogActivity(Activity4v2))
8.  Skill Sends Message Activity Log(Activity5) => Bot logs incoming activity Log(Activity5)

Every time we are logging an activity we are creating the blob and setting metadata on the blob.  Because of the size of the trace activities this turns into 3 storage writes X 13 transaction log writes which we are awaiting all of them.

It was taking almost 3-5 seconds for a skill to respond, and it was all backed up in waiting for the blob writes.

## To Reproduce

1.  Set up a bot/skill environment with TranscriptMiddleware enabled on both sides.
2.  Add some trace activities to your skill
3.  Count the transactions which occure.

## Expected behavior

The original code was correct, we should not be awaiting the logging to transcript store.

In addition, we should make sure that outbound SkillAdapter code does not forward TraceActivities if the channel is not Emulator because it's really expensive.

Additionally, I observed that the legacy Azure Transcript Store is uploading 2 blocks (2 round trips) to write the blob.  Carlos and I looked into this and it doesn't look like you can control this with the old library, but with the new library there is a parameter to UploadBlob() that lets you set a large size so we can write the blob in 1 round trip.  We should make sure we are setting that.

## Screenshots

![](https://user-images.githubusercontent.com/17789481/93009134-c1180180-f531-11ea-850d-6c0de7889aba.png)

**Comment by gabog** [comment]
Hi @tomlm, what ITranscriptLogger implementation were you using when you saw the 3/5 secs bottleneck? 

![image](https://user-images.githubusercontent.com/12264946/93025632-68964200-f5cd-11ea-92ca-841881c5317a.png)

I agree that we shouldn't be logging skill activities that are rerouted to the adapter (and maybe traces) and created an issue for that a couple of days ago [here](https://github.com/microsoft/botframework-sdk/issues/6031)

But I am not sure if we need to drop the await, under load conditions we may start getting exception and dropping log entries if we just fire and forget. I added await Task .WhenAll() to allow for the logging calls to be done in parallel so the issue may be with the underlying ITranscriptLogger we are using and not with the middleware itself. 

We could solve the bottleneck by making a better ITranscriptLogger that is faster or add a background process for ITranscriptLogger or TranscriptLoggerMiddleware that ensure all log entries are written under load conditions. Not awaiting the call may be just hiding issue with the underlying logging class if it doesn't scale.

**Comment by gabog** [comment]
I made some comments for the PR, once that's addressed we can merge that, let us know if the fix needs to be cherry picked for 4.10.x.

**Comment by EricDahlvang** [comment]
> Additionally, I observed that the legacy Azure Transcript Store is uploading 2 blocks (2 round trips) to write the blob.  Carlos and I looked into this and it doesn't look like you can control this with the old library, but with the new library there is a parameter to UploadBlob() that lets you set a large size so we can write the blob in 1 round trip.  We should make sure we are setting that.

Tom and Carlos raise a good point.  We should consider what is the optimal `StorageTransferOptions`

Created a separate issue to investigate that: https://github.com/microsoft/botbuilder-dotnet/issues/4684
